client/main.go

```diff
@@ -465,6 +478,168 @@ func schemeUnderstood(scheme string) error {
 	return nil
 }
 
+// Function for the object ls command, we get target information for our remote object and eventually print out the contents of the specified object
+func DoList(ctx context.Context, remoteObject string, options ...TransferOption) (err error) {
+	// First, create a handler for any panics that occur
+	defer func() {
+		if r := recover(); r != nil {
+			log.Debugln("Panic captured while attempting to perform transfer (DoList):", r)
+			log.Debugln("Panic caused by the following", string(debug.Stack()))
+			ret := fmt.Sprintf("Unrecoverable error (panic) captured in DoList: %v", r)
+			err = errors.New(ret)
+		}
+	}()
+
+	remoteObjectUrl, err := url.Parse(remoteObject)
+	if err != nil {
+		return errors.Wrap(err, "failed to parse remote object URL")
+	}
+
+	remoteObjectScheme := remoteObjectUrl.Scheme
+
+	// Check if we understand the found url scheme
+	err = schemeUnderstood(remoteObjectScheme)
+	if err != nil {
+		return err
+	}
+	te := NewTransferEngine(ctx)
+	defer func() {
+		if err := te.Shutdown(); err != nil {
+			log.Errorln("Failure when shutting down transfer engine:", err)
+		}
+	}()
+
+	pelicanURL, err := te.newPelicanURL(remoteObjectUrl)
+	if err != nil {
+		return errors.Wrap(err, "Failed to generate pelicanURL object")
+	}
+
+	ns, err := getNamespaceInfo(ctx, remoteObjectUrl.Path, pelicanURL.directorUrl, false, "")
+	if err != nil {
+		return err
+	}
+
+	// Get our token if needed
+	tokenLocation := ""
+	acquire := true
+	token := ""
+	long := false
+	dironly := false
+	fileonly := false
+	JSON := false
+	for _, option := range options {
+		switch option.Ident() {
+		case identTransferOptionTokenLocation{}:
+			tokenLocation = option.Value().(string)
+		case identTransferOptionAcquireToken{}:
+			acquire = option.Value().(bool)
+		case identTransferOptionToken{}:
+			token = option.Value().(string)
+		case identTransferOptionLong{}:
+			long = option.Value().(bool)
+		case identTransferOptionDirOnly{}:
+			dironly = option.Value().(bool)
+		case identTransferOptionFileOnly{}:
+			fileonly = option.Value().(bool)
+		case identTransferOptionJson{}:
+			JSON = option.Value().(bool)
+		}
+	}
+
+	if ns.UseTokenOnRead && token == "" {
+		token, err = getToken(remoteObjectUrl, ns, true, "", tokenLocation, acquire)
+		if err != nil {
+			return errors.Errorf("failed to get token for transfer: %v", err)
+		}
+	}
+
+	fileInfos, err := listHttp(ctx, remoteObjectUrl, pelicanURL.directorUrl, ns, token, WithLongOption(long), WithDirOnlyOption(dironly), WithFileOnlyOption(fileonly))
+	if err != nil {
+		return err
+	}
+
+	// Take our fileInfos and print them in a nice way
+	// if the -L flag was set, we print more information
+	if long {
+		w := tabwriter.NewWriter(os.Stdout, 1, 2, 10, ' ', tabwriter.TabIndent|tabwriter.DiscardEmptyColumns)
+		// If we want JSON format, we append the file info to a slice of fileInfo structs so that we can marshal it
+		if JSON {
```

Comment:
Making a note for me here, I realized that JSON as a variable name is not good since it will be exported. Silly mistake on my end and needs to be changed